py/sentry_streams/pipeline.py

```diff
@@ -0,0 +1,84 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, MutableMapping, Optional
+
+from sentry_streams.adapters.stream_adapter import StreamAdapter
+
+
+class RuntimeTranslator:
+    def __init__(self, runtime_adapter: StreamAdapter):
+        self.adapter = runtime_adapter
+
+    def translate_step(self, step: Step, stream: Optional[Any] = None) -> Any:
+        assert hasattr(step, "step_type")
+        next_step = step.step_type
+        translated_fn = getattr(self.adapter, next_step)
```

Comment:
What about simplifying this for now and just do a bunch of instance of with exhaustive type checking so no new type can be missed?

```
from typing import assert_never

if isinstance(step, Sink):
        transalte sink
    elif isinstance(step, KafkaSource):
        translate source
    else:
        assert_never(widget.datapoint) # This fails if a new class is added that is not part of this if then else
```

Reply:
In order to use `assert_never` I have to do some thing like or some combo of:

```
if isinstance(step, KafkaSource):
            return self.adapter.source(step)

        elif isinstance(step, Sink):
            return self.adapter.sink(step, stream)

        elif isinstance(step, WithInput):
            pass

        elif isinstance(step, KafkaSink):
            pass

        elif isinstance(step, Step):
            pass
        else:
            assert_never(step)
```

Perhaps I can just use a RuntimeError for now?

Reply:
Also, a bigger point is that if we avoid using a static property on the dataclass to denote the step type and instead extend base classes (which generally makes more sense), then we end up with empty classes like this https://github.com/getsentry/streams/pull/8/files#diff-6e974549d7fef65165b6d06969fe9017e604a654169f1510052e5adce52fd135R59

There will likely be a few different Sinks, but Sinks are not much more than a `WithInput`.

Reply:
The empty class is actually a "pattern" called marker interface even if not very common in python https://en.wikipedia.org/wiki/Marker_interface_pattern

Though we can say that is generally a symptom of a system that is not properly modularized. At this point I would not stress out too much about this as we are likely going to make fundamental changes to this data structure very soon and being able to unblock parallel work is more important.

Reply:
Let's do this for this PR:
- create an enum with the types of steps we support (so the property is not a bare string that we can break)
- add the static property that contains a value of the enum above. 
- in the `translate_step` method enumerate the values of the enum and find the method to call that way. While a bit brittle at least we will be able to statically type check. Then we can make it smarter.
- you can use the `assert_never` to ensure the enum is checked exhaustively. 

Then let's add a few more primitives and then revisit the data structure.